tools/codegen/model_generator/model.go

```diff
@@ -0,0 +1,257 @@
+	if err != nil {
+		// This func should only get called for FileDescriptor that has this attribute,
+		// therefore it is impossible to get to this state.
+		return fmt.Errorf("file option %s is required", tmplExtns.E_TemplateVariety.Name)
+	}
+	if *(tmplVariety.(*tmplExtns.TemplateVariety)) == tmplExtns.TemplateVariety_TEMPLATE_VARIETY_CHECK {
+		m.Check = true
+		m.VarietyName = "Check"
+	} else {
+		m.Check = false
+		m.VarietyName = "Report"
+	}
+	return nil
+}
+
+// Build field information about the Constructor message.
+func (m *Model) addInstanceFieldFromConstructor(parser *FileDescriptorSetParser, fdp *FileDescriptor) error {
+	m.ConstructorFields = make([]FieldInfo, 0)
+	cstrDesc, err := getRequiredMsg(fdp, "Constructor")
+	if err != nil {
+		return err
+	}
+	for _, fieldDesc := range cstrDesc.Field {
+		fieldName := CamelCase(fieldDesc.GetName())
+		typename := parser.GoType(cstrDesc.DescriptorProto, fieldDesc)
+		// TODO : Can there be more than one expressions in a type for a field in Constructor ?
+		typename = strings.Replace(typename, FullNameOfExprMessageWithPtr, "interface{}", 1)
+
+		m.ConstructorFields = append(m.ConstructorFields, FieldInfo{Name: fieldName, Type: TypeInfo{Name: typename}})
+	}
+	return nil
+}
+
+func (m *Model) addImports(parser *FileDescriptorSetParser, fileDescriptor *FileDescriptor, typDesc *Descriptor) {
+	usedPackages := getReferencedPackagesWithinType(parser, typDesc)
+	m.Imports = make([]string, 0)
+	for _, s := range fileDescriptor.Dependency {
+		fd := parser.fileByName(s)
+		// Do not import our own package.
+		if fd.PackageName() == parser.packageName {
+			continue
+		}
+		filename := fd.goFileName()
+		// By default, import path is the dirname of the Go filename.
+		importPath := path.Dir(filename)
+		if substitution, ok := parser.ImportMap[s]; ok {
+			importPath = substitution
+		}
+
+		pname := goPackageName(fd.GetPackage())
+		if !contains(usedPackages, pname) {
+			pname = "_"
+		}
+		m.Imports = append(m.Imports, pname+" "+strconv.Quote(importPath))
+	}
+}
+
+func getRequiredMsg(fdp *FileDescriptor, msgName string) (*Descriptor, error) {
+	var cstrDesc *Descriptor = nil
+	for _, desc := range fdp.desc {
+		if desc.GetName() == msgName {
+			cstrDesc = desc
+			break
+		}
+	}
+	if cstrDesc == nil {
+		return nil, fmt.Errorf("%s should have a message '%s'", fdp.GetName(), msgName)
+	}
+	return cstrDesc, nil
+}
+
+func getReferencedPackagesWithinType(parser *FileDescriptorSetParser, typDesc *Descriptor) []string {
+	result := make([]string, 0)
+	for _, fieldDesc := range typDesc.GetField() {
+		getReferencedPackagesByField(parser, fieldDesc, typDesc.PackageName(), &result)
+	}
+
+	return result
+}
+
+func getReferencedPackagesByField(parser *FileDescriptorSetParser, fieldDesc *descriptor.FieldDescriptorProto, parentPackage string, referencedPkgs *[]string) {
+	if *fieldDesc.Type == descriptor.FieldDescriptorProto_TYPE_MESSAGE || *fieldDesc.Type == descriptor.FieldDescriptorProto_TYPE_ENUM {
```

Comment:
nit: rewrite where the `if` clause just wraps a return and the rest of the method can be written without the extra level of indent.